cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,387 @@
+			cs := fmt.Sprintf("Component %s failed install:", cn)
+			l.logAndPrintf("\n%s\n%s\n", cs, strings.Repeat("=", len(cs)))
+			l.logAndPrint("Error: ", out[cn].Err, "\n")
+		} else {
+			cs := fmt.Sprintf("Component %s installed successfully:", cn)
+			l.logAndPrintf("\n%s\n%s\n", cs, strings.Repeat("=", len(cs)))
+		}
+
+		if strings.TrimSpace(out[cn].Stderr) != "" {
+			l.logAndPrint("Error detail:\n", out[cn].Stderr, "\n")
+		}
+		if strings.TrimSpace(out[cn].Stdout) != "" {
+			l.logAndPrint("Stdout:\n", out[cn].Stdout, "\n")
+		}
+	}
+}
+
+func getKubeClient(kubeConfig, configContext string) kubernetes.ExecClient {
+	kubeClient, err := clientExecFactory(kubeConfig, configContext)
+	if err != nil {
+		l.logAndFatalf("Abort. Failed to connect Kubernetes API server: %v", err)
+	}
+	return kubeClient
+}
+
+func checkUpgradeValues(curValues string, tarValues string, yes bool) {
+	diff := compare.YAMLCmp(curValues, tarValues)
+	if diff == "" {
+		l.logAndPrintf("Upgrade check: values are valid for upgrade.\n")
+	} else {
+		l.logAndPrintf("Upgrade check: values will be changed during the upgrade:\n%s", diff)
+	}
+
+	if yes {
+		return
+	}
+	if !confirm("Confirm to proceed [y/N]?", os.Stdout) {
+		l.logAndFatalf("Abort.")
+	}
+}
+
+func getImageSourceOverlay(v string) string {
+	setImageSource := []string{
+		"hub=docker.io/istio",
+		"tag=" + v,
+	}
+	imageSourceOverlay, err := makeTreeFromSetList(setImageSource)
+	if err != nil {
+		l.logAndFatal(err.Error())
+	}
+	return imageSourceOverlay
+}
+
+func genValuesFromFile(targetVer, filename string) string {
+	imageSourceOverlay := getImageSourceOverlay(targetVer)
+	values, err := genProfile(true, filename, "", imageSourceOverlay, "")
+	if err != nil {
+		l.logAndFatalf("Abort. Failed to generate values from file: %v, error: %v", filename, err)
+	}
+	return values
+}
+
+func clientExecFactory(kubeconfig, configContext string) (kubernetes.ExecClient, error) {
+	return kubernetes.NewClient(kubeconfig, configContext)
+}
+
+func readValuesFromInjectorConfigMap(kubeClient kubernetes.ExecClient, istioNamespace string) string {
+	configMapList, err := kubeClient.ConfigMapForSelector(istioNamespace, "istio=sidecar-injector")
+	if err != nil || len(configMapList.Items) == 0 {
+		l.logAndFatalf("Abort. Failed to retrieve sidecar-injector config map: %v", err)
+	}
+
+	jsonValues := ""
+	foundValues := false
+	for _, item := range configMapList.Items {
+		if item.Name == "istio-sidecar-injector" && item.Data != nil {
+			jsonValues, foundValues = item.Data["values"]
+			if foundValues {
+				break
+			}
+		}
+	}
+
+	if !foundValues {
+		l.logAndFatalf("Abort. Failed to find values in sidecar-injector config map: %v", configMapList)
+	}
+
+	yamlValues, err := yaml.JSONToYAML([]byte(jsonValues))
+	if err != nil {
+		l.logAndFatalf("jsonToYAML failed to parse values:\n%v\nError:\n%v", yamlValues, err)
+	}
+
+	return string(yamlValues)
+}
+
+func checkSupportedVersions(cur string, tar string) {
```

Comment:
all of this is supported through the go-version library included through version/version.go. at least we should not need parseVersionFormat.

Reply:
Fixed. Using the go-version library included through version/version.go instead.